pkg/scraper/client/resource/client.go

```diff
@@ -0,0 +1,164 @@
+	useNodeStatusPort bool
+	client            *http.Client
+	scheme            string
+	addrResolver      utils.NodeAddressResolver
+	buffers           sync.Pool
+}
+
+func NewClient(config client.KubeletClientConfig) (*kubeletClient, error) {
+	transport, err := rest.TransportFor(&config.Client)
+	if err != nil {
+		return nil, fmt.Errorf("unable to construct transport: %v", err)
+	}
+
+	c := &http.Client{
+		Transport: transport,
+	}
+	return &kubeletClient{
+		addrResolver:      utils.NewPriorityNodeAddressResolver(config.AddressTypePriority),
+		defaultPort:       config.DefaultPort,
+		client:            c,
+		scheme:            config.Scheme,
+		useNodeStatusPort: config.UseNodeStatusPort,
+		buffers: sync.Pool{
+			New: func() interface{} {
+				return new(bytes.Buffer)
+			},
+		},
+	}, nil
+}
+
+var _ client.KubeletMetricsInterface = (*kubeletClient)(nil)
+
+type ErrNotFound struct {
+	endpoint string
+}
+
+func (err *ErrNotFound) Error() string {
+	return fmt.Sprintf("%q not found", err.endpoint)
+}
+
+func (kc *kubeletClient) GetMetrics(ctx context.Context, node *corev1.Node) (*storage.MetricsBatch, error) {
+	port := kc.defaultPort
+	nodeStatusPort := int(node.Status.DaemonEndpoints.KubeletEndpoint.Port)
+	if kc.useNodeStatusPort && nodeStatusPort != 0 {
+		port = nodeStatusPort
+	}
+	addr, err := kc.addrResolver.NodeAddress(node)
+	if err != nil {
+		return nil, fmt.Errorf("unable to extract connection information for node %q: %v", node.Name, err)
+	}
+	url := url.URL{
+		Scheme: kc.scheme,
+		Host:   net.JoinHostPort(addr, strconv.Itoa(port)),
+		Path:   "/metrics/resource",
+	}
+
+	req, err := http.NewRequest("GET", url.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	client := kc.client
+	if client == nil {
+		client = http.DefaultClient
+	}
+	resource, err := kc.makeRequest(client, req.WithContext(ctx))
+	if err != nil {
+		return nil, err
+	}
+	samples, err := decodeProm004(resource)
+	if err != nil {
+		return nil, err
+	}
+	return decodeBatch(samples, node.Name), err
+}
+
+func decodeProm004(resource string) ([]*model.Sample, error) {
+	dec := expfmt.NewDecoder(strings.NewReader(resource), expfmt.FmtText)
+	decoder := expfmt.SampleDecoder{
+		Dec:  dec,
+		Opts: &expfmt.DecodeOptions{},
+	}
+
+	var samples []*model.Sample
+	for {
+		var v model.Vector
+		if err := decoder.Decode(&v); err != nil {
+			if err == io.EOF {
+				// Expected loop termination condition.
+				return samples, nil
+			}
+			return nil, err
+		}
+		samples = append(samples, v...)
+	}
+}
+
+func (kc *kubeletClient) makeRequest(client *http.Client, req *http.Request) (string, error) {
+	response, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer response.Body.Close()
+	b := kc.getBuffer()
+	defer kc.returnBuffer(b)
+	_, err = io.Copy(b, response.Body)
+	if err != nil {
+		return "", err
+	}
+	if response.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("request failed - %q.", response.Status)
+	}
+	return b.String(), nil
```

Comment:
This like copies content from buffer to string. This invalidates resource saving of reusing same buffer. Please decode data directly from buffer to avoid copying.